InmemoryTable filter for key value from query

spuniani opened the pull request:
When a query is passed into ReadAll, InMemoryTable assumes that it is of the type "PRIMARY_KEY_ID = ?" with the value in whereArguments. It retrieves the first argument and then uses it to get an object from the dataMap which is then passed back to the calling function.

If query is null or if there are no arguments in the where clause, then just return all elements from the Map.

abhideep commented on src/main/java/com/sarality/db/InMemoryTable.java:
```diff
@@ -41,8 +40,27 @@ public Long create(T data) {
     List<T> dataList = new ArrayList<>();
-    dataList.addAll(dataMap.values());
+
+    //null query, just return everything
```
nitpic: space after //

spuniani replied:
done

abhideep commented on src/main/java/com/sarality/db/InMemoryTable.java:
```diff
@@ -41,8 +40,27 @@ public Long create(T data) {
+    if (whereArgs.length != 0) {
+      Long key = Long.parseLong(whereArgs[0]);
+      dataList.add(dataMap.get(key));
+    }
```
formatting 

if () {
  yada yada
} else {
  more yada
}

abhideep commented on src/main/java/com/sarality/db/InMemoryTable.java:
```diff
@@ -41,8 +40,27 @@ public Long create(T data) {
+    }
+    else {
+      //return everything: no other type of query is supported
+      dataList.addAll(dataMap.values());
```
Throw an exception. Take a look at comments in the CL that you closed. We should just make sure that we know about the PrimaryKey Column and just check for a where clause of type "PRIMARY_KEY_COLUMN = ?". If nothing else we should have a TODO for that

spuniani replied:
added a method that classes implementing InMemoryTable have to implement to return the pk column. Used that to compare with the column name and throw an exception

abhideep commented on src/main/java/com/sarality/db/InMemoryTable.java:
```diff
@@ -41,8 +40,27 @@ public Long create(T data) {
+    String[] whereArgs = query.getWhereClauseArguments();
+
+    //assumes that there is only 1 argument in the where clause which is a primary key Id to retrieve.
+    if (whereArgs.length != 0) {
```
Check for length = 1 and throw Unsupported or IllegalArgument exception otherwise

spuniani replied:
done

spuniani commented:
Moved the query manipulation into a SimpleQueryExtractor which does the reverse of SimpleQueryBuilder and then provides lists of the column names, operator names and arguments to manipulate. 

moved the key extraction code into a private method as it will also be used by update.

abhideep commented:
Extractors job is to extract. The business logic for whether InMemoryTable supports this type of query does not reside with the Extractor.

This is a far more complicated route than what I was talking about, but since it is mostly implemented we can just go that route, though that too still needs a lot of work.

Here are your two options.

1) You check if the query.whereClause is equal to getPrimaryKey.getName() + " = ?", if so we support it, other we don't and throw an exception. That's the simple one.

2) The other is to create a class say ParsedQuery with a static constructor parseFrom(Query query), just like the class below. This creates strongly types Lists of Columns operators and arguments, not strings that you have below. After that you perform the same test as 1) but with strongly types values and you still do them in the InMemoryTable

spuniani commented:
I restructured the code as we discussed. I added a TODO to strongly type the query parser Lists - for now it just returns strings.

I also added a couple of methods - a check for supported queries and a getKey method that retrieves the key from the argument list.

abhideep commented on src/main/java/com/sarality/db/InMemoryTable.java:
```diff
@@ -18,9 +19,10 @@
-
   public abstract void setId(T data, Long id);
 
+  public abstract Column getPrimaryKeyField();
```
Can we rename to getPrimaryKeyColumn or just getPrimaryKey

abhideep commented on src/main/java/com/sarality/db/InMemoryTable.java:
```diff
@@ -41,8 +43,27 @@ public Long create(T data) {
+    }
+
+    // parse the query
+    QueryParser queryParser = new QueryParser(query);
```
This is a ParsedQuery rather than a parser.

abhideep commented on src/main/java/com/sarality/db/InMemoryTable.java:
```diff
@@ -41,8 +43,27 @@ public Long create(T data) {
+    QueryParser queryParser = new QueryParser(query);
+
+    // check if this is a supported query
+    if (!isSupportedQueryType(queryParser)) {
```
nitpic: let's have a method for each type of supported query. In this case the method is

isPrimaryKeyQuery(parsedQuery)

abhideep commented on src/main/java/com/sarality/db/query/QueryParser.java:
```diff
@@ -0,0 +1,56 @@
+/**
+ * Given a query, reverses the Query Builder operations and extracts the columnlist, operators and arguments
+ *
+ * @author satya@ Satya Puniani
```
We should look into formatting things properly somehow

The format here is 

@author satya@ (Satya Puniani)

abhideep commented on src/main/java/com/sarality/db/query/QueryParser.java:
```diff
@@ -0,0 +1,56 @@
+    String operatorNames = TextUtils.join("|",LogicalOperator.values());
+    String[] whereClause = query.getWhereClause().split(operatorNames);
+
+    //for each where clause, get columnname, operator, argument
```
Please add a TODO here to refine this further and throw ParseException when the class is not able to process the query

abhideep commented on src/main/java/com/sarality/db/query/QueryParser.java:
```diff
@@ -0,0 +1,56 @@
+ *
+ * @author satya@ Satya Puniani
+ */
+public class QueryParser {
```
Class name ParsedQuery

spuniani commented:
fixed issues from review. 
Update method now handles the query parameter to update the underlying object when called.
